### PR TITLE
IPC4: add ipc4 message support

### PR DESCRIPTION
**Pull request by RanderWang**
There are two types of ipc4 message: global message and module message.
    Ipc message is first decoded in handler.c and processed with the help
    of helper.c to deal with pipeline message and module message.
    Pipeline and Module messages are converted to sof pipeline and
    component message.
    
    Ipc4 pipeline message <------> ipc3 pipeline message
    RUNNING       <------->   TRIGGER START
    INIT + PAUSED <------->   PIPELINE COMPLETE
    PAUSED        <------->   TRIGER_PAUSE
    RESET         <------->   TRIGER_STOP + RESET
    EOS           <------->   TRIGER_RELEASE
    
    Ipc4 Module message <------> ipc3 module message
    init module             <------->   create component
    bind modules            <------->   connect components
    module set_large_config <------->   component cmd
    delete module           <------->   free component
    
    There is no buffer module in ipc4 design and FW manages buffer for each
    module while buffer component is defined by topology and create by ipc
    message for ipc3 design. For ipc4 path, FW will create buffer when two
    modules are bound since the buffer format can be gathered from source
    & sink modules and the buffer can be connected to both modules.
    
    Currently only part of ipc4 messages are processed and others will be done
    in future.

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@RanderWang this looks like a code move only. Can you confirm.

**Comment by RanderWang** [comment]
> @RanderWang this looks like a code move only. Can you confirm.

@lgirdwood  please check ipc4/helper.c and ipc4/handler.c. you may need to load diff

File: src/include/sof/ipc/topology.h [hunk 1]
```diff
@@ -45,6 +45,12 @@ typedef uint32_t ipc_comp;
 #include <ipc4/gateway.h>
 #define ipc_from_pipe_new(x) ((struct ipc4_pipeline_create *)x)
 #define ipc_from_pipe_connect(x) ((struct ipc4_module_bind_unbind *)x)
+
+const struct comp_driver *ipc4_get_comp_drv(int module_id);
+struct comp_dev *ipc4_get_comp_dev(uint32_t comp_id);
+int ipc4_add_comp_dev(struct comp_dev *dev);
+const struct comp_driver *ipc4_get_drv(uint8_t *uuid);
+int ipc_comp_disconnect(struct ipc *ipc, ipc_pipe_comp_connect *_connect);
```
Comment (lgirdwood) [inline src/include/sof/ipc/topology.h:53]: Looks like ```ipc_comp_disconnect()``` is generic API and not a ipc4 API. This should be moved to be with the generic IPC APIs

File: src/ipc/ipc4/handler.c [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+	ipc_cmd_hdr *hdr = ipc_get()->comp_data;
+	return hdr;
+}
+
+/*
+ * Most ABI 4 messages use compact format - keep logic simpler
+ * and handle all in IPC command.
+ */
+static uint32_t msg_in[2]; /* local copy of current message from host header */
+static uint32_t msg_out[2]; /* local copy of current message to host header */
```
Comment (lgirdwood) [inline src/ipc/ipc4/handler.c:439]: Lets do something better here. We should really have a structure for this, even if it's a structure that contains 2 uint32_t

**Review by lgirdwood — COMMENTED** [review]
Ok, the first patch is mostly a code move except for changing the calling parameters of one function which is fine.

File: src/include/sof/ipc/topology.h (hunk 1 above)
Comment (RanderWang) [reply src/include/sof/ipc/topology.h:53]: sure

File: src/ipc/ipc4/helper.c [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+	/* create the pipeline */
+	pipe = pipeline_new(pipe_desc->header.r.instance_id,
+			    pipe_desc->header.r.ppl_priority, 0);
+	if (!pipe) {
+		tr_err(&ipc_tr, "ipc: pipeline_new() failed");
+		return IPC4_OUT_OF_MEMORY;
+	}
+
+	pipe->time_domain = SOF_TIME_DOMAIN_TIMER;
+	/* 1ms */
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/ipc/ipc4/helper.c:113]: sure

File: src/ipc/ipc4/handler.c (hunk 2 above)
Comment (RanderWang) [reply src/ipc/ipc4/handler.c:439]: sure

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@RanderWang can you fix the conflicts. Thanks !

**Comment by RanderWang** [comment]
fixed conflict and cmocka

**Comment by RanderWang** [comment]
> @RanderWang can you check CI
> 
> ```
> 5:36:01.196   => Exception in test.
> 15:36:01.212     Exception DspPanicError occurred at 'C:\\work\\tools\\pyexecnetcache\\cavs_scripts-py\\utilities\\ipc_driver.py':599
>                  DSP Panic on IPC response timeout!
> ```
> 
> This is in IPC3 mode.
> 
> Fwiw, you could split this into 2 PRs to help root cause the CI failure.

thanks, fixed it.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
CI showing known and unrelated Zephyr issues.

**Pull request merged**